File: java/vm/src/test/java/simplevm/vm/SimpleTests.java
package simplevm.vm;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import static simplevm.vm.Bytecode.*;

class SimpleTests {
    @Test void testTest() {
        assertEquals(2, 1 + 1);
        //assertEquals(2, 1 - 1); // just to fail-test, just in case
    }
    @Test void testInstantiation() {
        VirtualMachine vm = new VirtualMachine();

        assertTrue(vm != null);
    }
    @Test void testNOP() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(NOP);

        // If we got here, with no exception, we're good
        assertTrue(true);
    }
    @Test void testNOPs() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            NOP,
            NOP,
            NOP,
            NOP
        });

        // If we got here, with no exception, we're good
        assertTrue(true);
    }
    @Test void testDump() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(DUMP);

        // If we got here, with no exception, we're good
        assertTrue(true);
    }
    @Test void testTrace() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            TRACE,
            NOP,
            DUMP
        });

        // If we got here, with no exception, we're good
        assertEquals(0, vm.getStack().length);
    }
    @Test void testPrint() {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            CONST, 12,
            PRINT
        });

        // Check the output for the printed output

        assertEquals(0, vm.getStack().length);
    }
}

File: cs/vmtest/SimpleTests.cs
namespace vmtest;

using vm;

[TestClass]
public class SimpleTests
{
    [TestMethod]
    public void TestTest()
    {
        Assert.AreEqual(2, 1 + 1);
        //Assert.AreEqual(2, 1 - 1); // just to make sure tests can fail, just in case
    }

    [TestMethod]
    public void TestInstantiation()
    {
        VirtualMachine vm = new VirtualMachine();

        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestNOP()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(Bytecode.NOP);

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestNOPs()
    {
        VirtualMachine vm = new VirtualMachine();

        Bytecode[] code = {
            Bytecode.NOP,
            Bytecode.NOP,
            Bytecode.NOP,
            Bytecode.NOP
        };
        vm.Execute(code);

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }

    [TestMethod]
    public void TestPRINT()
    {
        VirtualMachine vm = new VirtualMachine();
        
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)12,
            Bytecode.PRINT
        });
        
        // Check out the output window to see if it printed

        Assert.AreEqual(0, vm.Stack.Length);
    }

    [TestMethod]
    public void TestDUMP()
    {
        VirtualMachine vm = new VirtualMachine();

        Bytecode[] code = {
            Bytecode.NOP,
            Bytecode.DUMP
        };
        vm.Execute(code);

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }
    [TestMethod]
    public void TestTRACE()
    {
        VirtualMachine vm = new VirtualMachine();

        Bytecode[] code = {
            Bytecode.TRACE,
            Bytecode.NOP,
            Bytecode.DUMP
        };
        vm.Execute(code);

        // If we get here, we're good
        Assert.IsTrue(vm != null);
    }
}
